x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+	}
+
+	store.Delete(types.GetGVGsBindingOnBucketKey(bucketID))
+	return nil
+}
+
+func (k Keeper) BindingEmptyObjectToGVG(ctx sdk.Context, bucketID math.Uint, primarySPID, familyID uint32) (*types.LocalVirtualGroup, error) {
+	family, found := k.GetGVGFamily(ctx, primarySPID, familyID)
+	if !found {
+		return nil, types.ErrGVGFamilyNotExist
+	}
+
+	if len(family.GlobalVirtualGroupIds) == 0 {
+		return nil, types.ErrGVGNotExist
+	}
+
+	gvgID := family.GlobalVirtualGroupIds[0]
+
+	return k.BindingObjectToGVG(ctx, bucketID, primarySPID, familyID, gvgID, 0)
+}
+
+func (k Keeper) SwapOutAsPrimarySP(ctx sdk.Context, primarySP, successorSP *sptypes.StorageProvider, familyID uint32) error {
+	store := ctx.KVStore(k.storeKey)
+
+	family, found := k.GetGVGFamily(ctx, primarySP.Id, familyID)
+	if !found {
+		return types.ErrGVGFamilyNotExist
+	}
+	var gvgs []*types.GlobalVirtualGroup
+	for _, gvgID := range family.GlobalVirtualGroupIds {
+		gvg, found := k.GetGVG(ctx, gvgID)
+		if !found {
+			return types.ErrGVGNotExist
+		}
+		if gvg.PrimarySpId != primarySP.Id {
+			return types.ErrSwapOutFailed.Wrapf(
+				"the primary id (%d) in global virtual group is not match the primary sp id (%d)", gvg.PrimarySpId, primarySP.Id)
+		}
+		for _, spID := range gvg.SecondarySpIds {
+			if spID == successorSP.Id {
+				return types.ErrSwapOutFailed.Wrapf("the successor primary sp(ID: %d) can not be the secondary sp of gvg(%s).", successorSP.Id, gvg.String())
+			}
+		}
+
+		// swap deposit
+		if !gvg.TotalDeposit.IsZero() {
+			// send back deposit
+			coins := sdk.NewCoins(sdk.NewCoin(k.DepositDenomForGVG(ctx), gvg.TotalDeposit))
+			err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sdk.MustAccAddressFromHex(primarySP.FundingAddress), coins)
+			if err != nil {
+				return err
+			}
+
+			// successor deposit
+			err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, sdk.MustAccAddressFromHex(successorSP.FundingAddress), types.ModuleName, coins)
+			if err != nil {
+				return err
+			}
+		}
+
+		gvg.PrimarySpId = successorSP.Id
+		gvgs = append(gvgs, gvg)
+	}
+
+	// settlement
+	err := k.SettleAndDistributeGVGFamily(ctx, primarySP.Id, family)
+	if err != nil {
+		return types.ErrSwapOutFailed.Wrapf("fail to settle GVG family %d", familyID)
+	}
+
+	k.SetGVGFamily(ctx, successorSP.Id, family)
+	for _, gvg := range gvgs {
+		k.SetGVG(ctx, gvg)
+	}
+	store.Delete(types.GetGVGFamilyKey(primarySP.Id, familyID))
+	return nil
+}
+
+func (k Keeper) SwapOutAsSecondarySP(ctx sdk.Context, secondarySPID, successorSPID uint32, gvgIDs []uint32) error {
+	for _, gvgID := range gvgIDs {
+		gvg, found := k.GetGVG(ctx, gvgID)
+		if !found {
+			return types.ErrGVGNotExist
+		}
+		if gvg.PrimarySpId == successorSPID {
+			return types.ErrSwapOutFailed.Wrapf("the successor primary sp(ID: %d) can not be the primary sp of gvg(%s).", successorSPID, gvg.String())
+		}
+
+		// settlement
+		err := k.SettleAndDistributeGVG(ctx, gvg)
```

Comment:
we should check `successorSPID` is not with the secondrary sps in the gvgs.

Reply:
Fixed.